source/server/http/admin.cc

```diff
@@ -514,6 +514,29 @@ Http::Code AdminImpl::handlerCpuProfiler(absl::string_view url, Http::HeaderMap&
   return Http::Code::OK;
 }
 
+Http::Code AdminImpl::handlerHeapProfiler(absl::string_view url, Http::HeaderMap&,
+                                          Buffer::Instance& response, AdminStream&) {
+  Http::Utility::QueryParams query_params = Http::Utility::parseQueryString(url);
+  if (query_params.size() != 1 || query_params.begin()->first != "enable" ||
+      (query_params.begin()->second != "y" && query_params.begin()->second != "n")) {
+    response.add("?enable=<y|n>\n");
+    return Http::Code::BadRequest;
+  }
+
+  bool enable = query_params.begin()->second == "y";
+  if (enable && !Profiler::Heap::profilerEnabled()) {
+    if (!Profiler::Heap::startProfiler(profile_path_)) {
+      response.add("failure to start the heap profiler");
```

Comment:
It'd be better to produce a different error message if profiling has already been started than if profiling is not available.

Reply:
I wanted to make the error message rather simple. There are quite a lots of failure reasons including dumping file path not available. But I think you are right on whether the build is supporting this feature.

Reply:
Yeah I have a strong preference for actionable error messages. So I'd vote for propagating these out.

Reply:
Now return 5XX if profiler is not available in build and 4XX if profiler has already started. Also the response body will be different.